akka-remote/src/main/scala/akka/remote/artery/AeronSink.scala

```diff
@@ -28,26 +29,35 @@ import org.agrona.concurrent.UnsafeBuffer
 
 object AeronSink {
 
-  class OfferTask(pub: Publication, var buffer: UnsafeBuffer, msgSize: AtomicInteger, onOfferSuccess: AsyncCallback[Unit])
-    extends (() ⇒ Boolean) {
+  final class GaveUpSendingException(msg: String) extends RuntimeException(msg) with NoStackTrace
 
+  final class OfferTask(pub: Publication, var buffer: UnsafeBuffer, var msgSize: Int, onOfferSuccess: AsyncCallback[Unit],
+                        giveUpAfter: Duration, onGiveUp: AsyncCallback[Unit])
+    extends (() ⇒ Boolean) {
+    val giveUpAfterNanos = giveUpAfter match {
+      case f: FiniteDuration ⇒ f.toNanos
+      case _                 ⇒ -1L
+    }
     var n = 0L
-    var localMsgSize = -1
+    var startTime = 0L
 
     override def apply(): Boolean = {
+      if (n == 0L) {
+        // first invocation for this message
+        startTime = if (giveUpAfterNanos >= 0) System.nanoTime() else 0L
+      }
       n += 1
-      if (localMsgSize == -1)
-        localMsgSize = msgSize.get
-      val result = pub.offer(buffer, 0, localMsgSize)
+      val result = pub.offer(buffer, 0, msgSize)
       if (result >= 0) {
-        n = 0
-        localMsgSize = -1
+        n = 0L
         onOfferSuccess.invoke(())
         true
+      } else if (giveUpAfterNanos >= 0 && n % 10000 == 0 && (System.nanoTime() - startTime) > giveUpAfterNanos) {
```

Comment:
don't do the division, it will be run for every iteration. Do it every 8192 iteration, which is basically `&`-ing the last 13 bits and check if zero.